### PR TITLE
optimize WeightNorm on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51692 optimize WeightNorm on CPU**

### CODE REVIEW

**Comment by mingfeima** [comment]
This PR aims at improving `weight_norm` performance on CPU.
CUDA path has a fused kernel when `dim=0` or `dim=-1`, aka. the first dimension or the last dimension. This PR did pretty much the same thing on the CPU side.

Performance benchmarking is done on both avx2 CPU and avx512 CPU:
* **avx2 CPU**: Xeon(R) CPU E5-2680 v3 @ 2.50GHz, dual sockets, 12 cores per socket. Codename **Haswell**, [spec](https://ark.intel.com/content/www/us/en/ark/products/81908/intel-xeon-processor-e5-2680-v3-30m-cache-2-50-ghz.html)
*  **avx512 CPU**: Xeon(R) Gold 6248 CPU @ 2.50GHz, dual sockets, 20 cores per socket. Codename **Cascade Lake**, [spec](https://ark.intel.com/content/www/us/en/ark/products/192446/intel-xeon-gold-6248-processor-27-5m-cache-2-50-ghz.html)

benchmark script placed at [weight_norm.py](https://github.com/mingfeima/op_bench-py/blob/master/weight_norm.py)
PS. I think it is pointless to measure `weight_norm` itself, so I measured `torch.nn.utils.weight_norm(Linear)`. Thus each iteration will actually be `weight_norm_forward + linear(gemm)`, so the actual performance gain of `weight_norm_forward` part will be greater than what is shown in the following tables.

unit: ms per interation.

#### AVX2 CPU: batch_size=1: single core inference

dim | batch_size | input_channels | output_channels | before | after | speedup
-- | -- | -- | -- | -- | -- | --
first | 1 | 256 | 256 | 0.295 | 0.140 | 2.11
first | 1 | 256 | 512 | 0.494 | 0.165 | 2.99
first | 1 | 512 | 512 | 0.777 | 0.216 | 3.60
first | 1 | 512 | 1024 | 1.469 | 0.379 | 3.88
first | 1 | 1024 | 1024 | 2.533 | 0.737 | 3.44
last | 1 | 256 | 256 | 0.292 | 0.099 | 2.95
last | 1 | 256 | 512 | 0.492 | 0.152 | 3.24
last | 1 | 512 | 512 | 0.778 | 0.215 | 3.62
last | 1 | 512 | 1024 | 1.471 | 0.381 | 3.86
last | 1 | 1024 | 1024 | 2.521 | 0.685 | 3.68

#### AVX2 CPU: batch_size=128: single socket inference

dim | batch_size | input_channels | output_channels | before | after | speedup
-- | -- | -- | -- | -- | -- | --
first | 128 | 256 | 256 | 0.292 | 0.158 | 1.85
first | 128 | 256 | 512 | 0.359 | 0.163 | 2.20
first | 128 | 512 | 512 | 0.414 | 0.236 | 1.75
first | 128 | 512 | 1024 | 0.657 | 0.373 | 1.76
first | 128 | 1024 | 1024 | 0.933 | 0.602 | 1.55
last | 128 | 256 | 256 | 0.265 | 0.118 | 2.25
last | 128 | 256 | 512 | 0.356 | 0.159 | 2.24
last | 128 | 512 | 512 | 0.428 | 0.225 | 1.90
last | 128 | 512 | 1024 | 0.659 | 0.377 | 1.75
last | 128 | 1024 | 1024 | 0.923 | 0.612 | 1.51

#### AVX512 CPU: batch_size=1: single core inference

dim | batch_size | input_channels | output_channels | before | after | speedup
-- | -- | -- | -- | -- | -- | --
first | 1 | 256 | 256 | 0.190 | 0.057 | 3.33
first | 1 | 256 | 512 | 0.341 | 0.107 | 3.19
first | 1 | 512 | 512 | 0.561 | 0.175 | 3.21
first | 1 | 512 | 1024 | 1.068 | 0.330 | 3.24
first | 1 | 1024 | 1024 | 1.852 | 0.538 | 3.44
last | 1 | 256 | 256 | 0.182 | 0.057 | 3.19
last | 1 | 256 | 512 | 0.335 | 0.108 | 3.10
last | 1 | 512 | 512 | 0.560 | 0.175 | 3.20
last | 1 | 512 | 1024 | 1.063 | 0.338 | 3.14
last | 1 | 1024 | 1024 | 1.860 | 0.544 | 3.42

#### AVX512 CPU: batch_size=128: single socket inference

dim | batch_size | input_channels | output_channels | before | after | speedup
-- | -- | -- | -- | -- | -- | --
first | 128 | 256 | 256 | 0.172 | 0.072 | 2.39
first | 128 | 256 | 512 | 0.226 | 0.092 | 2.46
first | 128 | 512 | 512 | 0.258 | 0.127 | 2.03
first | 128 | 512 | 1024 | 0.380 | 0.184 | 2.07
first | 128 | 1024 | 1024 | 0.545 | 0.342 | 1.59
last | 128 | 256 | 256 | 0.172 | 0.072 | 2.39
last | 128 | 256 | 512 | 0.223 | 0.091 | 2.45
last | 128 | 512 | 512 | 0.257 | 0.127 | 2.02
last | 128 | 512 | 1024 | 0.381 | 0.184 | 2.07
last | 128 | 1024 | 1024 | 0.545 | 0.345 | 1.58

**Comment by mingfeima** [comment]
@VitalyFedyunin @ngimel could you please review this one?

**Comment by VitalyFedyunin** [comment]
Hello! I will try to allocate some time next week for this PR. Can you please rebase by Monday. Thanks.

File: aten/src/ATen/native/cpu/WeightNormKernel.cpp [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+      //
+      // since we are doing vectorization on N, abstract grad_w and v:
+      //   grad_v = a * grad_w + b * v
+      //       a := g / norms
+      //       b := -per_dim_sums * g / (norms^3)
+      //
+      scalar_t g_val = g_data[m];
+      scalar_t norms_val = norms_data[m];
+      scalar_t norms3_val = norms_val * norms_val * norms_val;
+      scalar_t a = g_val / norms_val;
```
Comment (glaringlee) [inline aten/src/ATen/native/cpu/WeightNormKernel.cpp:168]: Is that possible norms_val == 0 ? Do we need to handle this case?

File: aten/src/ATen/native/cpu/WeightNormKernel.cpp [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+      per_dim_sums_data[n] = per_dim_sums_val;
+
+      // fusing parallel session is marginally faster.
+      // update parameters
+      //   a := g / norms
+      //   b := -per_dim_sums * g / (norms^3)
+      scalar_t g_val = g_data[n];
+      scalar_t norms_val = norms_data[n];
+      scalar_t norms3_val = norms_val * norms_val * norms_val;
+      a_data[n] = g_val / norms_val;
```
Comment (glaringlee) [inline aten/src/ATen/native/cpu/WeightNormKernel.cpp:253]: Same here, `norms_val == 0`

**Review by glaringlee — COMMENTED** [review]
Also this will break the CI backward compatibility test since the following api (name) changed
_weight_norm_cuda_interface_backward
_weight_norm_cuda_interface
You need to add above two functions into the allow list here with a future date (say 1month ahead):
https://github.com/pytorch/pytorch/blob/master/test/backward_compatibility/check_backward_compatibility.py

And since the above 2 apis' name changed. You have to update more places, here is a search for _weight_norm_cuda_interface_backward
https://github.com/pytorch/pytorch/search?q=_weight_norm_cuda_interface_backward
You can do the same thing to the other one, update all places that used those 2 apis with the new names.
